### ISSUE TITLE
[FIREBASE_ADMOB] Two simultaneous listeners not working

### ISSUE THREAD

**Issue by hasen6** [FirebaseExtended/flutterfire#135]
**Describe the bug**
With a listener attached to both an interstitial and a banner, only one listener is 'listened' to.

**To Reproduce**
Steps to reproduce the behaviour:
1. Add both an interstitial and a banner with listeners on each.
2. Printed events are only for one ad type

**Expected behaviour**
I would expect different outputs for each listener so as to act accordingly on the responses.

```
  InterstitialAd interstitialAd;
  BannerAd bannerAd;
 
  BannerAd myBanner() {
    return BannerAd(
      adUnitId: 'ca-app-pub-...',
      size: AdSize.smartBanner,
      targetingInfo: targetingInfo,
      listener: (MobileAdEvent event) {
        print("BannerAd event is $event");
        if (event == MobileAdEvent.failedToLoad) {
          bannerAd..load();
        } else if (event == MobileAdEvent.loaded) {
          bannerAd
            ..show(
              anchorOffset: 0.0,
              anchorType: AnchorType.bottom,
            );
        }
      },
    );
  }

  InterstitialAd myInterstitial() {
    return InterstitialAd(
      adUnitId: 'ca-app-pub-...',
      targetingInfo: targetingInfo,
      listener: (MobileAdEvent event) {
        print("InterstitialAd event is $event");
        if (event == MobileAdEvent.failedToLoad) {
          interstitialAd..load();
        } else if (event == MobileAdEvent.closed) {
          interstitialAd = myInterstitial()..load();
        }
      },
    );
  }

    @override
  void initState() {
    super.initState();
    bannerAd = myBanner()..load();
    interstitialAd = myInterstitial()..load();
  }

```

All I get is reports from the banner ad since it is initialised first in `initState`, BUT it is clear that the events are for both the banner AND the interstitial, even though they are from one listener. I want the interstitial to reload when it is closed and the banner ad to show when it is loaded. 

Maybe I've done something incorrectly?

**Comment by BondarenkoStas** [comment]
Duplicate of #63
@hasen6 Thank you for your code example. I'm closing the issue as duplicate, reference on it will be in the older issue.

**Comment by hasen6** [comment]
That issue seems different, is not that the InterstitialAd listener is not working, but rather that they can't function at the same time. Both ad listeners work perfectly fine, banner and interstitial, as long as one is disabled.

**Comment by crossplatfrom** [comment]
Can confirm but only for iOS. Android is working with multiple listeners for me

**Comment by ThrowJojo** [comment]
Also to add to this(not sure if these two issues are connected?), it seems like if you have an interstitial and a banner and you dispose one of them with `.dispose()` it also disposes the other - even though they have different IDs. I think there might be an issue with how IDs are being handled in the iOS code.

https://github.com/FirebaseExtended/flutterfire/blob/master/packages/firebase_admob/ios/Classes/FirebaseAdMobPlugin.m#L235

https://github.com/FirebaseExtended/flutterfire/blob/master/packages/firebase_admob/ios/Classes/FLTMobileAd.m#L89

**Comment by JCKodel** [comment]
Same here, only for iOS.
My scenario:
I have one fixed banner on the bottom of the app, shown 100% of the time (except when the user subscribes to a in app purchase, in this case, I `.dispose()` the banner.

Additionally, I have interstitial and rewarded videos that are displayed when the users haven't enough virtual coins.

There is a Firebase Remote Config that chooses between those two, so 66% shows interstitial, 33% rewarded videos.

Then, I expect two events in the interstitial/rewarded videos:
1) For interstitial, I give users some coins when the event is `impression` (this only works on iOS) or `opened` (this only works on Android) and `leftApplication` gives them extra coins (because they clicked the ad)
2) For rewarded videos, `rewarded` and `leftApplication`

Only for iOS, no interstitial events are dispatched, so, basically, it never works (if I force them to show, it appears, once (because my reload is in the `closed` event, which never happens). If I try to redisplay it, it says something like "this already been displayed, so, won't display it again" (there is an actual error message in console, from "google").

**Comment by ThrowJojo** [comment]
Maybe this issue is related somehow to https://github.com/FirebaseExtended/flutterfire/issues/1328 ?

**Comment by crossplatfrom** [comment]
This issue is still unresolved as of the date of this comment

**Comment by ThrowJojo** [comment]
Yeah it still hasn't been resolved. It's really just a matter of somebody fixing the class variable issues described in #1328. Myself and a couple of others made some small changes after forking and it seemed to function as intended. I would have a crack at it but like I said in #1328 I'm not that confident with Objective-C and I think the reason this issue exists in the first place is because somebody without extensive knowledge of Objective-C created the plugin. When I first noticed the issue I looked at the source code and couldn't understand what was wrong until I read @ouxiaoyong wrote weeks later in #1328.

**Comment by orestesgaolin** [comment]
I have similar issue that on iOS the smart banner ad is not shown if medium rectangle was shown before.

**Comment by AR553** [comment]
every time i wanted to call the interstitial ad i call like this:

interstitialAd = createInterstitialAd()..load();
interstitialAd..load()..show();

and this fixed the error of showing ad only once in the widget's lifetime.
Hope it helps someone.

**Comment by Logic2Paradigm** [comment]
That's what worked for me (in Flutter):

`
    import 'package:firebase_admob/firebase_admob.dart';
    
    InterstitialAd _interstitialAd;
    bool _isInterstitialAdReady;
    
    void prepAd(){
      initState();
      _loadInterstitialAd();
    }
    
    void runAd() {
      _interstitialAd.show();
    }
    
    void initState() {
      _isInterstitialAdReady = false;
      _interstitialAd = InterstitialAd(
        adUnitId: InterstitialAd.testAdUnitId,
        listener: _onInterstitialAdEvent,
      );
    }
    
    
    void _loadInterstitialAd() {
      _interstitialAd.load();
    }
    // TODO: Implement _onInterstitialAdEvent()
    void _onInterstitialAdEvent(MobileAdEvent event) {
      switch (event) {
        case MobileAdEvent.loaded:
          _isInterstitialAdReady = true;
          break;
        case MobileAdEvent.failedToLoad:
          _isInterstitialAdReady = false;
          print('Failed to load an interstitial ad');
          break;
        case MobileAdEvent.closed:
          dispose();
          prepAd();
          break;
        default:
        // do nothing
      }
    }
    
    void dispose() {// TODO: Dispose InterstitialAd object
      _interstitialAd?.dispose();
    }`

I ran prepAd() once in the beginning and then just runAd() when I need it. You can check _isInterstitialAdReady == true inside runAd() if you want

**Comment by russellwheatley** [comment]
The `firebase_admob` plugin will be deprecated in April 2021 in favor of [Google Mobile Ads SDK for Flutter](https://github.com/googleads/googleads-mobile-flutter).
Google Mobile Ads SDK for Flutter is a new Flutter plugin that supports more Ads formats than `firebase_admob` does. Google Mobile Ads SDK for Flutter currently supports loading and displaying banner, interstitial (full-screen), native ads, and rewarded video ads across AdMob and AdManager. It also supports displaying banner and native ads as `Widgets` as opposed to being overlayed over all app content.
We're now recommending projects currently using `firebase_admob` migrate to Google Mobile Ads SDK for Flutter following the [instructions outlined here](https://github.com/FirebaseExtended/flutterfire/tree/master/packages/firebase_admob#migrating-to-google-mobile-ads-sdk-for-flutter).
Issues with migrating or with using the new package should now be raised over on the [Google Mobile Ads SDK for Flutter](https://github.com/googleads/googleads-mobile-flutter) repository.

**Comment by JCKodel** [comment]
> The `firebase_admob` plugin will be deprecated in April 2021 in favor of [Google Mobile Ads SDK for Flutter](https://github.com/googleads/googleads-mobile-flutter).
> Google Mobile Ads SDK for Flutter is a new Flutter plugin that supports more Ads formats than `firebase_admob` does. Google Mobile Ads SDK for Flutter currently supports loading and displaying banner, interstitial (full-screen), native ads, and rewarded video ads across AdMob and AdManager. It also supports displaying banner and native ads as `Widgets` as opposed to being overlayed over all app content.
> We're now recommending projects currently using `firebase_admob` migrate to Google Mobile Ads SDK for Flutter following the [instructions outlined here](https://github.com/FirebaseExtended/flutterfire/tree/master/packages/firebase_admob#migrating-to-google-mobile-ads-sdk-for-flutter).
> Issues with migrating or with using the new package should now be raised over on the [Google Mobile Ads SDK for Flutter](https://github.com/googleads/googleads-mobile-flutter) repository.

That plugin is a joke. Most events won't even work (such as AdClosed).

Guess we'll have to earn money through mopub then...
